### PR TITLE
Create a follow up report on echis if there is a mismatch of client details during registration

### LINKED ISSUE #117: Flag mismatching client details on eCHIS for CHP Follow up
> Upon submitting the same national ID with different names, the two clients get the same UPI. Ensure that the mismatching details are flagged for the CHP to follow up instead of updating with the UPI.

Issue from https://github.com/moh-kenya/config-echis-2.0/issues/1353

### PR DESCRIPTION
**Pull request by freddieptf**
Fixes #117 
Ready for review @kitsao @karimkkanji, you can test this on [stagin](https://chis-staging.health.go.ke). Next step is working on the chp follow up task in echis

### CODE REVIEW

**Comment by freddieptf** [comment]
Follow up task has been worked on in https://github.com/moh-kenya/config-echis-2.0/pull/1394, also up on staging for testing

**Comment by freddieptf** [comment]
@alexosugo wanna have a look here? Reviews welcome

**Comment by freddieptf** [comment]
Ping @kitsao @karimkkanji

File: src/controllers/client.js [hunk 1]
```diff
@@ -115,24 +115,36 @@ const updateDocWithUPI = async (echisClientDoc, clientNumber) => {
...
+    };
+    const response = await axiosInstance.post(`api/v2/records`, body);
+    logger.information(COMPLETED_SUCCESSFULLY);
+    return response;
+
+  } catch (error) {
+    logger.error(error);
+    return error;
+  }
 };
```
Comment (kitsao) [inline src/controllers/client.js]: Still a lot ongoing here. We can retain it for now, but this is a good candidate for refactoring.

File: src/controllers/client.js [hunk 2]
```diff
@@ -130,15 +130,14 @@ const createClientDetailsMismatchReport = async (echisClientDoc, fields) => {
...
       },
       patient_id: echisClientDoc.doc._id,
-      authored_on: echisClientDoc.doc.reported_date,
-      mismatched_fields: fields,
+      mismatched_fields: fields.join(" "),
       source: `Client Registry`,
     };
     const response = await axiosInstance.post(`api/v2/records`, body);
-    logger.information(COMPLETED_SUCCESSFULLY);
+    logger.information(`COMPLETED ${response?.status}`);
```
Comment (kitsao) [inline src/controllers/client.js]: This needs to be explicit. Is it possible to complete unsuccessfully?

File: src/routes/client.js [hunk 3]
```diff
@@ -13,7 +13,7 @@ router.post("/", async function (req, res) {
     logger.information(CLIENT_ROUTE_COMPLETED);
 
     if (response?.upi || response?.msg === CLIENT_DETAILS_MISMATCH) {
-      logger.information(`${GENERATED_NUMBER} ${response.upi || response}`);
+      logger.information(`${response.upi || response.msg}`);
```
Comment (kitsao) [inline src/routes/client.js]: This is kinda confusing. Can we have `response?.msg === CLIENT_DETAILS_MISMATCH` with `response.upi` logged? I think we are better off moving that logic to a separate function and check explicitly.

File: src/controllers/referral.js [hunk 4]
```diff
@@ -25,24 +20,23 @@ const {
...
-
-  if (!upi) {
-    const echisDoc = await getEchisDocForUpdate(dataRecord._patient_id);
-    upi = echisDoc.upi;
+const getSubjectUpi = async (echisClientId) => {
+  try {
+    const echisClient = await getDoc({ instance: CHT.url, user: CHT.username, password: CHT.password }, echisClientId);
+    if (echisClient.upi) {
+      return echisClient.upi;
+    }
```
Comment (kitsao) [inline src/controllers/referral.js:29]: What should happen when there is no `upi`, but no error to be caught?

File: src/utils/client.js [hunk 5]
```diff
@@ -161,8 +167,24 @@ function getPropByString(obj, propString) {
...
+// compare fields in the echis client doc and the client doc we got from Client Registry
+function getMismatchedClientFields(echisClientDoc, crClientDoc) {
+  // fields that ideally should be kind of unique across clients
+  const matcherFields = ["firstName", "middleName", "lastName", "gender", "dateOfBirth", "contact.primaryPhone"];
+  // where we store our mismatched fields and return them later
+  const mismatchedFields = {};
+  // the payload we generate here has the same format as what we get back when we query for client existence
+  const crPayload = generateClientRegistryPayload(echisClientDoc);
+  matcherFields.forEach(key => {
+    if (getPropByString(crClientDoc, key).toUpperCase() !== getPropByString(crPayload, key).toUpperCase()) {
```
Comment (kitsao) [inline src/utils/client.js:179]: I see what you did here, but `toLowerCase()` would do be fine.

File: src/utils/client.js [hunk 6]
```diff
@@ -161,8 +167,24 @@ function getPropByString(obj, propString) {
...
+      mismatchedFields[key] = { Have: crPayload[key], Expected: crClientDoc[key] };
```
Comment (kitsao) [inline src/utils/client.js:180]: Any preference for `Have` and `Expected` to `have` and `expected`? `actual` vs `expected` sounds better - following common practice in _testing_.

File: src/utils/echis.js [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+
+function genRequestConfig(conf) {
+    return {
+        baseURL: conf.instance,
+        auth: {
+            username: conf.user,
+            password: conf.password,
+        },
+    }
+}
```
Comment (kitsao) [inline src/utils/echis.js:11]: Let us stick to the arrow notation for consistency.

**Review by kitsao — CHANGES_REQUESTED** [review]
Quite some heavy lifting you had to do here @freddieptf. The changes a generally good. I noted several areas that we could optimize and added comments inline. Note that some such as refactoring comments can be handled later.

File: src/controllers/client.js [hunk 8]
```diff
@@ -34,60 +34,44 @@ const axiosInstance = axios.create({
...
+      // update contact doc with UPI and return UPI
+      await updateDocWithUPI(echisClientDoc, response.data.client.clientNumber);
+      return {
+        upi: response.data.client.clientNumber,
+      };
     }
+    // if we get here, they did not exist in CR, we create them and update echis with the UPI
+    logger.information(CLIENT_NOT_FOUND);
+    logger.information(CREATE_IN_CR);
+    const createResponse = await createClientInRegistry(JSON.stringify(generateClientRegistryPayload(echisClientDoc)));
```
[the remark this answers is not in the document]
Comment (freddieptf) [reply src/controllers/client.js]: Already handled in https://github.com/moh-kenya/echis-2.0-integration/pull/118/commits/735239431d680e252ff67d824e28b4b0cef0dd89, make sure you're reviewing the correct change set

File: src/controllers/referral.js (hunk 4 above)
Comment (freddieptf) [reply src/controllers/referral.js:29]: We try and create the client in the next code block and if that fails then the function returns nothing

File: src/controllers/client.js [hunk 9]
```diff
@@ -1,270 +1,127 @@
...
+        res.status(200).send({
+          data: {
+            msg: CLIENT_DETAILS_MISMATCH,
+            fields: mismatchedFields
+          }
+        })
+      } catch (err) {
+        const errString = `could not create details mismatch report on ${CHT.url}: ${err.message}`
+        logger.error(errString)
+        res.status(400).send({ error: errString })
```
[the remark this answers is not in the document]
Comment (freddieptf) [reply src/controllers/client.js:97]: Yep, we have to rely on the err string. One thing is OpenHim will not retry a request if it returns a 5xx code so we have to be careful

File: src/utils/client.js (hunk 5 above)
Comment (freddieptf) [reply src/utils/client.js:179]: I was using the wisdom of the crowd, seems like `toUpperCase` works best for this https://stackoverflow.com/questions/2140627/how-to-do-case-insensitive-string-comparison#

File: src/controllers/client.js [hunk 10]
```diff
@@ -249,12 +233,20 @@ const updateEchisDocWithFailedIdentification = async (echisDoc) => {
...
+  const matcherFields = ["firstName", "middleName", "lastName", "gender", "dateOfBirth", "contact.primaryPhone"]
+  // where we store our mismatched fields and return them later
+  const mismatchedFields = [];
+  // the payload we generate here has the same format as what we get back when we query for client existence
+  const crPayload = generateClientRegistryPayload(echisContact);
+  matcherFields.forEach(key => {
+    if (getPropByString(crContact, key) !== getPropByString(crPayload, key)) {
+      mismatchedFields.push(key);
+    }
+  });
```
[the remark this answers is not in the document]
Comment (freddieptf) [reply src/controllers/client.js]: Yes we could do it that way but i chose this approach as it is more readable, at least to me.

File: src/utils/client.js [hunk 11]
```diff
@@ -142,7 +142,27 @@ const transformCountyCode = (code) => {
...
+    var candidate = obj[prop];
+    if (candidate !== undefined) {
+      obj = candidate;
+    } else {
+      break;
+    }
+  }
+  return obj[props[i]];
+}
+
```
[the remark this answers is not in the document]
Comment (freddieptf) [reply src/utils/client.js:169]: Oh i see, they're doing the same thing but there's no need to import a lib for one function

[the remark this answers is not in the document]
Comment (kitsao) [reply src/utils/client.js:169]: Just copy and paste :)

File: src/controllers/client.js (hunk 10 above)
[the remark this answers is not in the document]
Comment (kitsao) [reply src/controllers/client.js]: I realized that you later updated this anyway.

File: src/utils/client.js (hunk 11 above)
[the remark this answers is not in the document]
Comment (freddieptf) [reply src/utils/client.js:169]: Yes, i already did. It does not use `reduce` but works the exact same way

File: src/utils/client.js (hunk 6 above)
Comment (freddieptf) [reply src/utils/client.js]: done

File: src/utils/echis.js (hunk 7 above)
Comment (freddieptf) [reply src/utils/echis.js]: done

File: src/utils/client.js (hunk 11 above)
[the remark this answers is not in the document]
Comment (kitsao) [reply src/utils/client.js:169]: I meant you copy and paste the `getField` function here.

File: src/utils/client.js [hunk 12]
```diff
@@ -142,7 +142,27 @@ const transformCountyCode = (code) => {
...
+    var candidate = obj[prop];
+    if (candidate !== undefined) {
+      obj = candidate;
+    } else {
+      break;
+    }
+  }
+  return obj[props[i]];
+}
```
Comment (kitsao) [inline src/utils/client.js:168]: Use `let/const` for `var` and the arrow notation for consistency. I still think we use `getField`, copy and paste :)

**Review by kitsao — APPROVED** [review]
This should be OK. Merge after addressing the inline comment.

**Comment by karimkkanji** [comment]
@freddieptf 
image

This should not bar storing the UPI. Let's only check for Gender, names and date of birth

**Pull request merged**